feat: container API docs (#8281)
withastro/docs: src/content/docs/en/reference/container-reference.mdx

---
title: Astro Container API (experimental)
i18nReady: false
---
import Since from '~/components/Since.astro'

The Container API allows to render Astro components in isolation.

<Since v="4.9.0" />

## `create()`

Creates a new instance of the container. 

```js
import { experimental_AstroContainer } from "astro/container";

const container = await experimental_AstroContainer.create();
```

It accepts an object with the following options

```ts
export type AstroContainerOptions = {
	streaming?: boolean;
	renderers?: AstroRenderer[];
	astroConfig?: AstroContainerUserConfig;
};
```

#### `streaming` option

Enables rendering components using the streaming.

> **Type:** `boolean`

Author: Yeah that's correct

#### `renderers` option

A list of client renderers. Add a renderer if your Astro component renders some client components such as React, Preact, etc.

**Type:**: `AstroRenderer[]`;

> For example, if you're rendering a Astro component that renders a React component, you'll have to provide the following renderer: 

Author: Since we typed the options with `AstroRenderer`, users will be required to provide `name`, a `clientEntrypoint` renderer and `serverEntrypoint` renderer. 

Unfortunately, this is a very low-level API, and no end-user is aware of that other than developers who create Astro renderers. Renders that we support don't expose or document these things. 

Ideally, the example should provide *enough* knowledge as a starting point, but we plan to expose these renderers to the users, so they won't need to manually write them like in the example.

Overall, I think the suggestion looks

> For example, if you're rendering a Astro component that renders a React component, you'll have to provide the following renderer: 

Reviewer: (NOTE: Part of my thinking by exposing the vocab in a sentence is that even if they can't find these things themselves, it's easy enough for them to come into support and ask, "I can't find the `clientEntrypoint` and `serverEntrypoint` for Svelte..." and then that's an easy question for them to ask and get a direct answer to by someone who knows.)

```js
const container = await experimental_AstroContainer.create({
    renderers: [
      {
	      name: "@astrojs/react",
	      client: "@astrojs/react/client.js",
	      server: "@astrojs/react/server.js"
      }
    ]
})
const result = await container.renderToString(ReactWrapper);
```

#### `astroConfig` option

A subset of the Astro user configuration.

> **Type**: `AstroContainerUserConfig`

Author: We decided to remove this option from the container for the time being. We will wait for users to request specific options if it makes sense. We've had hard times to find a use case that made sense.

### Rendering options

Both [`renderToResponse`](#rendertoresponse) and [`renderToString`](#rendertostring) accept an object as their second argument:

```ts
export type ContainerRenderOptions = {
	slots?: Record<string, any>;
	request?: Request;
	params?: Record<string, string | undefined>;
	locals?: App.Locals;
	routeType?: "page" | "endpoint";
  
};
```

#### `slots`

**Type**: `Record<string, any>`;

Use this options if your component needs to render some [slots](en/basics/astro-components/#slots).

If your components renders one slot, pass an object with `default` as key:

```js name="Card.test.js"
import Card from "../src/components/Card.astro";

const result = await container.renderToString(Card, { 
  slots: { default: "Some value"}
});
```

If your component renders named slots, use the keys of the object to name the slots:

```astro name="Card.astro"
---
---
<div>
  <slot name="header" />
  <slot name="footer" />
</div>
```

```js name="Card.test.js"
import Card from "../src/components/Card.astro";

const result = await container.renderToString(Card, { 
  slots: { "header": "Header content", "footer": "Footer" }
});
```

> You can also render components in cascade:

Reviewer: I'm not sure we'd say "in cascade" but I get what you mean. Bookmarking this!

```astro name="Card.astro"
---
---
<div>
  <slot name="header" />
  <slot name="footer" />
</div>
```

```js name="Card.test.js"
import Card from "../src/components/Card.astro";
import CardHeader from "../src/components/CardHeader.astro";
import CardFooter from "../src/components/CardFooter.astro";

const result = await container.renderToString(Card, { 
  slots: { 
    "header": await container.renderToString(CardHeader), 
    "footer":  await container.renderToString(CardFooter),
  }
});
```

#### `request` option

**Type**: `Request`

> The request is used to understand which path/URL the component is about to render.
> 
> Use this option in case your component needs to read information like `Astro.url` or `Astro.request`.

Reviewer: ```suggestion
An option to pass a `Request` with information about the path/URL the component will render.

Use this option when your component needs to read information like `Astro.url` or `Astro.request`.
```

You can also inject possible headers or cookies.

```js file="Card.test.js"
import Card from "../src/components/Card.astro";

const result = await container.renderToString(Card, { 
  request: new Request("https://example.com/blog", {
    headers: {
      "X-some-secret-header": "test-value"
    }
  })
});
```

#### `params` option

**Type**: `Record<string, string | undefined>`;

> Use this option in case your component needs to read `Astro.params`. As opposed to `getStaticPaths`, you must provide only **one** item of the params returned by that function.

Reviewer: ```suggestion
An object to pass information about the path parameter to an Astro component responsible for [generating dynamic routes](/en/guides/routing/#dynamic-routes).

Use this option when your component needs a value for `Astro.params` in order to generate a single route dynamically.
```
Check this for accuracy, but I think this seems like what you're saying. 

And, I thought mentioning `getStaticPaths()` *might* be more confusing here than helpful? I think we can say this like above to emphasize that, like in other examples shown, you're providing *ONE* value of something so that you can test a particular case by rendering a single version of the component.

```astro name="Card.astro"
---
const { locale, slug } = Astro.params;
---
<div></div>
```

```js file="Card.test.js"
import Card from "../src/components/Card.astro";

const result = await container.renderToString(Card, { 
  params: {
    locale: "en",
    slug: "getting-started"
  }
});
```

#### `locals` options 

**Type**: `App.Locals`

> Use this option to stub the `Astro.locals` object:

Reviewer: ```suggestion
An option to pass information from [`Astro.locals`](/en/reference/api-reference/#astrolocals) for rendering your component.

Use this option to when your component needs information stored during the lifecycle of a request in order to render, such as logged in status.
```
These words might be kind of wrong, but just trying to fit this into the pattern!

```astro name="Card.astro"
---
const { checkAuth } = Astro.locals;
const isAuthenticated = checkAuth();
---
{isAuthenticated ? <span>You're in</span> : <span>You're out</span> }
```

```js file="Card.test.js"
import Card from "../src/components/Card.astro";

test("User is in", async () => {
  const result = await container.renderToString(Card, { 
    locals: {
      checkAuth() { return true }
    }
  });
  
  // assert result contains "You're in"
})

test("User is out", async () => {
  const result = await container.renderToString(Card, { 
    locals: {
      checkAuth() { return false }
    }
  });
  
  // assert result contains "You're out"
})
```

#### `routeType` option

**Type**: `"page" | "endpoint"`

> Useful in case you're rendering an endpoint. In this case, you want to use `renderToResponse`:

Reviewer: ```suggestion
An option available when using `renderToResponse` to specify that you are rendering an [endpoint](/en/guides/endpoints/):
```

```js
container.renderToString(Endpoint, { routeType: "endpoint" });
```

```js file="endpoint.test.js"
import * as Endpoint from "../src/pages/api/endpoint.js";

const response = await container.renderToResponse(Endpoint, { 
  routeType: "endpoint"
});
const json = await response.json();
```

> If your endpoint needs to be tested on methods such as `POST`, `PATCH`, etc., you'll have to use the `request` option to signal Astro to call the correct function:

Reviewer: ```suggestion
To test your endpoint on methods such as `POST`, `PATCH`, etc., use the `request` option to call the correct function:
```

import CardFooter from "../src/components/CardFooter.astro";

const result = await container.renderToString(Card, { 
  slots: { 
    "header": await container.renderToString(CardHeader), 
    "footer":  await container.renderToString(CardFooter),
  }
});
```

#### `request` option

**Type**: `Request`

An option to pass a `Request` with information about the path/URL the component will render.

Use this option when your component needs to read information like `Astro.url` or `Astro.request`.

You can also inject possible headers or cookies.

```js file="Card.test.js"
import Card from "../src/components/Card.astro";

const result = await container.renderToString(Card, { 
  request: new Request("https://example.com/blog", {
    headers: {
      "X-some-secret-header": "test-value"
    }
  })
});
```

#### `params` option

**Type**: `Record<string, string | undefined>`;

An object to pass information about the path parameter to an Astro component responsible for [generating dynamic routes](/en/guides/routing/#dynamic-routes).

Use this option when your component needs a value for `Astro.params` in order to generate a single route dynamically.

> ```astro name="Card.astro"

Author: Yeah, I think it makes more sense to use a file with a dynamic route
